### PR TITLE
Ntgl 402 and 393- custom colours/tutor cards dragging

### PR DESCRIPTION
**Pull request by M7s7**
Custom course colours
- Users can now customize the colours of each course via a colour-picker when editing a course
- Course colours are saved locally and are independent between different timetables
- Save button added to ExpandedEventView to facilitate colour picker confirmation

Tutorials
- Tutorials can no longer be dragged to invalid time slots
- Tutorial information cannot be edited by user
- Colour of tutorial can be changed on card menu
- Disabled event link in tutoring and lab events
- Created a new type, EventSubtype = 'General' | 'Tutoring', to distinguish between tutoring and custom events. 

UI/UX
- Colour picker button UI improved
- Cool line added

Authored-by: @jasttran and @M7s7

### CODE REVIEW

**Review by Shzmj — APPROVED** [review]
everything looks good to me

File: client/src/utils/cardsContextMenu.ts [hunk 1]
```diff
@@ -57,7 +57,17 @@ export const handlePasteEvent = (
...
+
+export const handleDeleteEvent = (
+  createdEvents: CreatedEvents,
+  setCreatedEvents: (createdEvents: CreatedEvents) => void,
+  eventPeriod: EventPeriod
+) => {
+  const updatedEventData = { ...createdEvents };
+  delete updatedEventData[eventPeriod.event.id];
+  setCreatedEvents(updatedEventData);
+};
```
Comment (Rayahhhmed) [inline client/src/utils/cardsContextMenu.ts:73]: add a new line here

File: client/src/App.tsx [hunk 2]
```diff
@@ -110,11 +110,31 @@ const App: React.FC = () => {
...
+    assignedColors,
+    setAssignedColors,
+  } = useContext(CourseContext);
 
   setDropzoneRange(days.length, earliestStartTime, latestEndTime);
 
+  /**
+   * Supports migration from non-assigned colors to assignedColors so timetable does not break. Can be removed after a few weeks when everyone has migrated
+   */
+  if (assignedColors === undefined) {
```
Comment (Rayahhhmed) [inline client/src/App.tsx:137]: Why strong typing here, is `assignedColors` a string that can be `''` or any other falsey value, then we have a flaw in our type system. Instead just rely on the falsey nature of `''`.  Like:
`if (!assignedColors) {...`

File: client/src/App.tsx [hunk 3]
```diff
@@ -110,11 +110,31 @@ const App: React.FC = () => {
...
+    const colors = useColorMapper(
+      displayTimetables[selectedTimetable].selectedCourses.map((c) => c.code),
```
Comment (Rayahhhmed) [inline client/src/App.tsx]: [Naming convention alert]: I know `c` is course here, but be descriptive and actually put it with something to do with`course`. Being a bit more verbose will save reader time

File: client/src/App.tsx [hunk 4]
```diff
@@ -171,6 +191,14 @@ const App: React.FC = () => {
     if (year !== invalidYearFormat) fetchReliably(fetchCoursesList);
   }, [year]);
 
+  // Fetching the saved timetables from local storage
+  useEffect(() => {
+    const savedTimetables = storage.get('timetables');
```
Comment (Rayahhhmed) [inline client/src/App.tsx]: possible to give a type for savedTImetables?

File: client/src/App.tsx [hunk 5]
```diff
@@ -278,6 +302,15 @@ const App: React.FC = () => {
       setSelectedCourses(newSelectedCourses);
       setCourseData(newCourseData);
 
+      if (displayTimetables.length > 0) {
+        setAssignedColors(
+          useColorMapper(
+            newSelectedCourses.map((c) => c.code),
```
Comment (Rayahhhmed) [inline client/src/App.tsx]: Same here, dont use `c`.

File: client/src/components/timetable/DroppedEvent.tsx [hunk 6]
```diff
@@ -55,8 +59,7 @@ const DroppedEvent: React.FC<DroppedEventProps> = ({
   };
 
   const onDown = (eventDown: any) => {
-    if (eventDown.button === 2 || contextMenu) return;
-
+    if (eventDown.button === 2 || contextMenu || eventPeriod.subtype === 'Tutoring') return;
```
Comment (Rayahhhmed) [inline client/src/components/timetable/DroppedEvent.tsx]: What is 2 here? dont use magic numbers have a const variable describing this value here instead

File: client/src/components/timetable/DroppedEvent.tsx [hunk 7]
```diff
@@ -175,15 +178,35 @@ const DroppedEvent: React.FC<DroppedEventProps> = ({
...
+            setContextMenu={setContextMenu}
+            setPopupOpen={setPopupOpen}
+            setIsEditing={setIsEditing}
+            setCopiedEvent={setCopiedEvent}
+            copiedEvent={copiedEvent}
+          />
+        ) : (
+          <>
+            <StyledMenu
+              open={contextMenu != null}
```
Comment (Rayahhhmed) [inline client/src/components/timetable/DroppedEvent.tsx]: why is `contextMenu != null`, is it going to be falsey? then just using a boolean flag is fine. Other wise use `!==`

**Review by Rayahhhmed — APPROVED** [review]
Left some comments, alter then and merge it in when u see fit

File: client/src/App.tsx (hunk 2 above)
Comment (M7s7) [reply client/src/App.tsx:137]: This is a temp fix, so it will be removed anyway later. I think we keep the same in order not to break anything

**Pull request merged**